libraries/AP_Math/AP_Math.h

```diff
@@ -16,273 +18,323 @@
+    
+    const auto ang_180 = 180.f*unit_mod;
+    const auto ang_360 = 360.f*unit_mod;
+    auto res = std::fmod(angle + ang_180, ang_360);
+    if (res < 0.f) {
+        res += (decltype(res))ang_360;
+    }
+    res -= (decltype(res))ang_180;
+    return res;
+}
 
-/*
-  wrap an angle defined in radians to -PI ~ PI (equivalent to +- 180 degrees)
+/* 
+ * @brief: Constrains an euler angle to be within the range: 0 to 360 degrees
+ * The second parameter changes the units. Standard: 1 == degrees, 10 == dezi, 100 == centi ..
  */
-float wrap_PI(float angle_in_radians);
+template <class T>
+auto wrap_360(const T &angle, float unit_mod = 1) -> decltype(std::fmod(angle, 360.f*unit_mod)) {
+    static_assert(std::is_floating_point<T>::value || std::is_integral<T>::value, "ERROR - wrap_360(): template parameter not of type float or int\n");
+    
+    const auto ang_360 = 360.f*unit_mod;
+    auto res = std::fmod(angle, ang_360);
+    if (res < 0.f) {
+        res += (decltype(res))ang_360;
+    }
+    return res;
+}
 
 /*
-  wrap an angle defined in radians to the interval [0,2*PI)
+ * Wrap an angle in centi-degrees
  */
-float wrap_2PI(float angle);
+template <class T>
+auto wrap_360_cd(const T &angle) -> decltype(wrap_360(angle, 100.f)) {
+    return wrap_360(angle, 100.f);
+}
 
 /*
- * check if lat and lng match. Ignore altitude and options
+ * Wrap an angle in centi-degrees
  */
-bool locations_are_same(const struct Location &loc1, const struct Location &loc2);
+template <class T>
+auto wrap_180_cd(const T &angle) -> decltype(wrap_180(angle, 100.f)) {
+    return wrap_180(angle, 100.f);
+}
 
 /*
-  print a int32_t lat/long in decimal degrees
+  wrap an angle defined in radians to -PI ~ PI (equivalent to +- 180 degrees)
  */
-void print_latlon(AP_HAL::BetterStream *s, int32_t lat_or_lon);
-
-// Converts from WGS84 geodetic coordinates (lat, lon, height)
-// into WGS84 Earth Centered, Earth Fixed (ECEF) coordinates
-// (X, Y, Z)
-void wgsllh2ecef(const Vector3d &llh, Vector3d &ecef);
-
-// Converts from WGS84 Earth Centered, Earth Fixed (ECEF)
-// coordinates (X, Y, Z), into WHS84 geodetic
-// coordinates (lat, lon, height)
-void wgsecef2llh(const Vector3d &ecef, Vector3d &llh);
-
-// constrain a value
-// constrain a value
-static inline float constrain_float(float amt, float low, float high)
-{
-	// the check for NaN as a float prevents propogation of
-	// floating point errors through any function that uses
-	// constrain_float(). The normal float semantics already handle -Inf
-	// and +Inf
-	if (isnan(amt)) {
-		return (low+high)*0.5f;
-	}
-	return ((amt)<(low)?(low):((amt)>(high)?(high):(amt)));
-}
-// constrain a int16_t value
-static inline int16_t constrain_int16(int16_t amt, int16_t low, int16_t high) {
-	return ((amt)<(low)?(low):((amt)>(high)?(high):(amt)));
+template <class T>
+auto wrap_PI(const T &radian) -> decltype(std::fmod(radian + static_cast<float>(M_PI), static_cast<float>(M_2_PI))) {
```

Comment:
@lucasdemarchi 
Made a flaw, this is how it may look like. Alternative would be to make an individual function for PX4. However, I don't think it would be desired.